8294437: java/nio/channels/FileChannel tests slow on Windows

djelinski opened the pull request:
Please review this test-only change that improves the execution speed of a few FileChannel tests:
- Transfer2GPlus, Transfer4GBFile, and TransferTo6GBFile were modified to use sparse files. Their speed is now consistent across platforms, provided that the platform / filesystem supports sparse files.
- LargeMapTest was rewritten to use sparse files, and to more precisely target the issue it was written to detect. In my tests it still crashed with EXCEPTION_ACCESS_VIOLATION when JDK-8286637 was reverted.
- MapTest: repetition was removed from `testForce` method; I'm not sure what issues it was supposed to catch, but at far as I can tell, it was only triggering timeouts, see [JDK-8289526](https://bugs.openjdk.org/browse/JDK-8289526), [JDK-8224480](https://bugs.openjdk.org/browse/JDK-8224480)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294437](https://bugs.openjdk.org/browse/JDK-8294437): java/nio/channels/FileChannel tests slow on Windows


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [0f22fb54](https://git.openjdk.org/jdk/pull/10464/files/0f22fb5418c0af51029d7adddb2fefdfbded0ca3)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to [0f22fb54](https://git.openjdk.org/jdk/pull/10464/files/0f22fb5418c0af51029d7adddb2fefdfbded0ca3)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10464/head:pull/10464` \
`$ git checkout pull/10464`

Update a local copy of the PR: \
`$ git checkout pull/10464` \
`$ git pull https://git.openjdk.org/jdk pull/10464/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10464`

View PR using the GUI difftool: \
`$ git pr show -t 10464`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10464.diff">https://git.openjdk.org/jdk/pull/10464.diff</a>

</details>

AlanBateman commented on test/jdk/java/nio/channels/FileChannel/LargeMapTest.java:
```diff
@@ -38,42 +41,44 @@
+        ByteBuffer bb;
+        try (FileChannel fc = FileChannel.open(p, CREATE_NEW, SPARSE, WRITE)) {
+            fc.position(BASE);
+            Random r = new Random(System.nanoTime());
```
It might not matter here but we usually use RandomFactory to allow for reproducibility in the event of test failure.

djelinski replied:
Thanks for the hint! It shouldn't matter here, but I'll change it.

AlanBateman commented on test/jdk/java/nio/channels/FileChannel/MapTest.java:
```diff
@@ -56,21 +57,42 @@ public class MapTest {
-            out.println("Exceptions: OK");
+            t1 = System.nanoTime();
+            out.printf("Exceptions: done in %d ns (%d ms) %n",
+                    t1 - t0, TimeUnit.NANOSECONDS.toMillis(t1 - t0));
```
Are you planning to leave all these timestamps in the output? I can't tell if they are left over from your testing or intentional.

djelinski replied:
Yeah, I wanted to leave them here; they don't take up much space, and are very useful when trying to figure out why the tests are slow.

AlanBateman reviewed (commented):
Creating the files as sparse files make sense, we probably should have changed these tests a long time ago.

bplb commented:
Overall I don’t see any problems aside from the copyright years which I assume you will fix before integrating. It’s good that you removed `RandomAccessFile` in one place. It might be good to try to remove all `java.io` uses where possible, e.g., not `deleteOnExit()`. Also, instead of creating a temporary file and then deleting it, maybe something else could be done. For example, creating a temporary directory and then creating a file with a hard-coded name within it. Or using `Random` to create a file name and using the `REPLACE_EXISTING` open option. Another thought would be to add a method `createSparseTempFile()` or something to `lib/util/FileUtils`.

djelinski commented:
I went with adding `FileUtils.createSparseTempFile()`; the method implementation is still ugly, but at least it's not visible when looking at the tests.

I added one more modification to `Transfer2GPlus`; noticed that `testToWritableByteChannel` doesn't need to write to disk, and modified the code to take advantage of that; it's much faster now.

AlanBateman commented:
> I went with adding `FileUtils.createSparseTempFile()`; the method implementation is still ugly, but at least it's not visible when looking at the tests.

I would prefer if the FileChannel tests do not use this. The reason is that these are tests for FileChannel and should not be dependent on test infrastructure that it out of sight. It would be okay for non-FileChannel tests to use of course.

bplb commented on test/jdk/java/nio/channels/FileChannel/LargeMapTest.java:
```diff
@@ -38,42 +44,44 @@
  *          EXCEPTION_ACCESS_VIOLATION.
  * @requires vm.bits == 64
- * @run main/othervm/timeout=240 LargeMapTest
+ * @library /test/lib
```
Typically tests which use randomly generated values also have the tag
```
 * @key randomness
```

bplb commented on test/jdk/java/nio/channels/FileChannel/Transfer2GPlus.java:
```diff
@@ -115,28 +130,58 @@ private static void testToFileChannel(Path src, byte[] expected)
+             WritableByteChannel wbc = Channels.newChannel(os)){
+
+            long n;
+            if ((n = srcCh.transferTo(0, LENGTH, wbc)) < LENGTH)
```
The contract of `FileChannel::transferTo` does not guarantee that all bytes are transferred:

```An invocation of this method may or may not transfer all of the requested bytes [...]```.

bplb replied:
No, I think the docs should be left as they are.

djelinski replied:
Ack that. If that's okay with you, I'd rather not change this code here; it was always like that, and I haven't found any related failure reports. The tests continue to pass.

bplb commented on test/jdk/java/nio/channels/FileChannel/Transfer4GBFile.java:
```diff
@@ -128,46 +108,33 @@ public void xferTest05() throws Exception { // for bug 4638365
-                sinkChannel.close();
+        try (FileChannel sourceChannel = FileChannel.open(source, StandardOpenOption.READ);
+             FileChannel sinkChannel = FileChannel.open(sink, StandardOpenOption.WRITE)) {
+            long bytesWritten = sinkChannel.transferFrom(sourceChannel,
```
Same comment as for `FileChannel::transferTo`:

```An invocation of this method may or may not transfer all of the requested bytes```.

bplb commented on test/jdk/java/nio/channels/FileChannel/Transfer4GBFile.java:
```diff
@@ -51,72 +50,53 @@ public class Transfer4GBFile {
+        try (FileChannel sourceChannel = FileChannel.open(source,
+                     StandardOpenOption.READ);
+             FileChannel sinkChannel = FileChannel.open(sink,
+                     StandardOpenOption.WRITE)) {
```
Some weird indentation at lines 71 and 73. A static import of the `StandardOpenOption` enum onstants would help.

AlanBateman commented on test/jdk/java/nio/channels/FileChannel/TransferTo6GBFile.java:
```diff
@@ -52,15 +55,12 @@ public void xferTest08() throws Exception { // for bug 6253145
 
         out.println("  Writing large file...");
         long t0 = System.nanoTime();
+        FileChannel fc = FileChannel.open(file, READ, WRITE);
```
You can probably change this to use try-with-resources too and maybe "READ" can be dropped as this loo just writes to the file.

AlanBateman reviewed (approved):
I think this version looks okay and probably okay drop the loop from testForce. It's probably best to wait until @bplb gets to see the updated version as he has been doing most of the work in this area recently.

AlanBateman reviewed (approved):
Thanks for the update, I don't have any other comments. Will be good to get this change in as these tests are so slow on Windows right now.

bplb commented on test/jdk/java/nio/channels/FileChannel/TransferTo6GBFile.java:
```diff
@@ -52,105 +55,96 @@ public void xferTest08() throws Exception { // for bug 6253145
-                        throw new RuntimeException("Premature EOF!");
-                    nread += n;
-                }
+            // Setup looback connection and echo server
```
"loopack" -> "loopback"

djelinski commented:
Typo fixed.

Thanks for the reviews!

/integrate
